### PR TITLE
range_search will early terminate with range_search_k

### PR DESCRIPTION
**Pull request by alwayslove2013**

### CODE REVIEW

File: include/knowhere/index/index_node.h [hunk 1]
```diff
@@ -95,41 +94,65 @@ class IndexNode : public Object {
...
                     continue;
                 }
-                num_consecutive_over_radius = 0;
+                if (range_search_k > 0) {
+                    if (early_stop_further_bounds.size() < range_search_k) {
+                        early_stop_further_bounds.emplace(dist);
+                    } else {
+                        early_stop_further_bounds.pop();
+                        early_stop_further_bounds.emplace(dist);
+                        further_bound = early_stop_further_bounds.top();
```
Comment (zhengbuqian) [inline include/knowhere/index/index_node.h:152]: `further_bound` is not monotonically getting closer, is that by design?

File: include/knowhere/index/index_node.h [hunk 2]
```diff
@@ -95,41 +94,65 @@ class IndexNode : public Object {
...
             auto it = its[idx];
             size_t num_next = 0;
-            size_t num_consecutive_over_radius = 0;
+            size_t num_consecutive_over_further_bound = 0;
             while (it->HasNext()) {
                 auto [id, dist] = it->Next();
                 num_next++;
-                // too close
-                if (has_range_filter && (similarity_metric ? dist > range_filter : dist < range_filter)) {
+                if (has_closer_bound && too_close(dist)) {
```
Comment (zhengbuqian) [inline include/knowhere/index/index_node.h:135]: maybe capture `has_closer_bound` in `too_close` as well

File: include/knowhere/index/index_node.h [hunk 3]
```diff
@@ -95,41 +94,65 @@ class IndexNode : public Object {
...
+        LOG_KNOWHERE_DEBUG_ << "range_search_k: " << range_search_k;
+
+        const bool the_larger_the_closer = IsMetricType(base_cfg.metric_type.value(), metric::IP) ||
+                                           IsMetricType(base_cfg.metric_type.value(), metric::COSINE) ||
+                                           IsMetricType(base_cfg.metric_type.value(), metric::BM25);
+        auto is_first_closer = [&the_larger_the_closer](const float dist_1, const float dist_2) {
+            return the_larger_the_closer ? dist_1 > dist_2 : dist_1 < dist_2;
+        };
+        auto too_close = [&is_first_closer, &closer_bound](float dist) { return is_first_closer(dist, closer_bound); };
+        auto too_far = [&is_first_closer, &further_bound](float dist) { return !is_first_closer(dist, further_bound); };
```
Comment (zhengbuqian) [inline include/knowhere/index/index_node.h:114]: `same_or_too_far`

Comment (alwayslove2013) [reply include/knowhere/index/index_node.h:114]: nice

File: include/knowhere/index/index_node.h (hunk 2 above)
Comment (alwayslove2013) [reply include/knowhere/index/index_node.h:135]: maybe it is better to separate.

File: include/knowhere/index/index_node.h (hunk 1 above)
Comment (zhengbuqian) [reply include/knowhere/index/index_node.h:152]: it is indeed monotonically getting closer

**Pull request merged**